zebra-state/src/service.rs

```diff
@@ -803,6 +803,91 @@ impl ReadStateService {
     fn latest_non_finalized_state(&self) -> NonFinalizedState {
         self.non_finalized_state_receiver.cloned_watch_data()
     }
+
+    /// Check the contextual validity of a block in the best chain
+    #[cfg(feature = "getblocktemplate-rpcs")]
+    #[allow(clippy::unwrap_in_result)]
+    #[tracing::instrument(level = "debug", skip_all)]
+    fn check_best_chain_contextual_validity(
+        &self,
+        prepared: &PreparedBlock,
+    ) -> Result<(), crate::ValidateContextError> {
+        let latest_non_finalized_state = self.latest_non_finalized_state();
+
+        let Some(best_chain) = latest_non_finalized_state.best_chain() else {
+            Err(crate::ValidateContextError::NotReadyToBeValidated)?
+        };
+
+        if best_chain.non_finalized_tip_hash() != prepared.block.header.previous_block_hash {
+            Err(crate::ValidateContextError::NotReadyToBeValidated)?
+        }
```

Comment:
Should this be a "not a child of the best chain tip" error?

Reply:
It probably should've been but we're just re-using the regular `validate_and_commit` fn now, which makes more sense given that that's what `submitblock` will call.